[jit] kill _parameter_list

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26666 [jit] module dedupe
* **#27399 [jit] kill _parameter_list**
* #26665 [jit Python None should have its type inferred as NoneType

This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
[jit] kill _parameter_list

This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

zdevito commented on torch/jit/quantized.py:
```diff
@@ -325,17 +323,19 @@ def process_weights(ihhh, layer, suffix, dtype):
+        if dtype == torch.int8:
+            self._orig_weights_names = self._packed_weights_names
+
+        self._packed_weights = torch.jit.Attribute([getattr(self, weight) for weight in self._packed_weights_names], List[Tensor])
```
Do we need to have torch.jit.Attribute or does it work to put the types in `__attributes__`?

suo replied:
It needs to be an attribute with the current ScriptModule API, but not after my rewrite.

zdevito reviewed (approved):
Thanks for cleaning this up.

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)

suo committed:
Update on "[jit] kill _parameter_list"


This was devised in a time when we didn't have module attributes. They
are essentially just tensor lists, so represent them that way. This has
the additional benefit of making the RNN forward pass faster because we
effectively cache the flattened weights.

The only complication part is that someone may come along and do:
```
my_rnn_mod.w_ih_l0 = torch.nn.Parameter(...)
```

This means we need to override setattr to keep the flattened weights
cache up to date.

Differential Revision: [D17785658](https://our.internmc.facebook.com/intern/diff/D17785658)
